### DISCUSSION TITLE
GraphSAGE Link Prediciton and Directed Graphs

### DISCUSSION THREAD

**Discussion by CalmLogarithm** [stellargraph/stellargraph#1776] [General] [answered]
I'm implementing the [GraphSAGE link prediction example](https://stellargraph.readthedocs.io/en/stable/demos/link-prediction/graphsage-link-prediction.html) in my own code. When I get to the test_train_split I get the error: "**Exception has occurred: NetworkXNotImplemented** not implemented for directed type". My graph is a MultiDiGraph. What is the suggested best way to get around this issue? Convert my graph to an undirected form?

**Comment by huonw** [comment, marked as answer]
Great to hear from you!

Could you say a little bit more about what the problem is? The link prediction example linked doesn't use NetworkX on the surface, so it's slightly hard to understand how you're putting things together. Specifically, it'd be helpful to know:

- some more details about the code; for example, the `train_test_split` call along how any parameters passed to it are defined
- the full stack trace: it will help narrow down exactly what methods are being called and where the problem is popping u

**Reply by CalmLogarithm** [reply]
Initially, I build the graph in NetworkX as a MultiDiGraph. There's also a separate Pandas dataframe that stores all the feature vectors. Then I use StellarGraph.from_networkx to pull them together into a StellarGraph object.

So the train_test_split code is almost directly from the example:

>     edge_splitter_test = EdgeSplitter(sg)
>     G_test, edge_ids_test, edge_labels_test = edge_splitter_test.train_test_split(
>         p=0.1, method='global', keep_connected=True
>     )

The full stack trace:
> Traceback (most recent call last):
>   File "C:\Users\court_000\AppData\Local\Programs\Python\Python36\lib\runpy.py", line 193, in _run_module_as_main
>     "__main__", mod_spec)
>   File "C:\Users\court_000\AppData\Local\Programs\Python\Python36\lib\runpy.py", line 85, in _run_code
>     exec(code, run_globals)
>   File "c:\Users\court_000\.vscode\extensions\ms-python.python-2020.6.91350\pythonFiles\lib\python\debugpy\__main__.py", line 45, in <module>
>     cli.main()
>   File "c:\Users\court_000\.vscode\extensions\ms-python.python-2020.6.91350\pythonFiles\lib\python\debugpy/..\debugpy\server\cli.py", line 430, in main
>     run()
>   File "c:\Users\court_000\.vscode\extensions\ms-python.python-2020.6.91350\pythonFiles\lib\python\debugpy/..\debugpy\server\cli.py", line 267, in run_file
>     runpy.run_path(options.target, run_name=compat.force_str("__main__"))
>   File "C:\Users\court_000\AppData\Local\Programs\Python\Python36\lib\runpy.py", line 263, in run_path
>     pkg_name=pkg_name, script_name=fname)
>   File "C:\Users\court_000\AppData\Local\Programs\Python\Python36\lib\runpy.py", line 96, in _run_module_code
>     mod_name, mod_spec, pkg_name, script_name)
>   File "C:\Users\court_000\AppData\Local\Programs\Python\Python36\lib\runpy.py", line 85, in _run_code
>     exec(code, run_globals)
>   File "c:\Users\court_000\Documents\Git\anansi\src\main\python\link_prediction.py", line 41, in <module>
>     main('data/eng/networkx.gpickle', 'xyz.csv')
>   File "c:\Users\court_000\Documents\Git\anansi\src\main\python\link_prediction.py", line 24, in main
>     p=0.1, method='global', keep_connected=True
>   File "C:\Users\court_000\AppData\Local\Programs\Python\Python36\lib\site-packages\stellargraph\data\edge_splitter.py", line 378, in train_test_split 
>     p=p, method=method, probs=probs, keep_connected=keep_connected
>   File "C:\Users\court_000\AppData\Local\Programs\Python\Python36\lib\site-packages\stellargraph\data\edge_splitter.py", line 114, in _train_test_split_homogeneous
>     self.minedges = self._get_minimum_spanning_edges()
>   File "C:\Users\court_000\AppData\Local\Programs\Python\Python36\lib\site-packages\stellargraph\data\edge_splitter.py", line 1077, in _get_minimum_spanning_edges
>     mst = nx.minimum_spanning_edges(self.g, data=False)
>   File "<decorator-gen-544>", line 2, in minimum_spanning_edges
>   File "C:\Users\court_000\AppData\Local\Programs\Python\Python36\lib\site-packages\networkx\utils\decorators.py", line 80, in _not_implemented_for    
>     raise nx.NetworkXNotImplemented(msg)
> networkx.exception.NetworkXNotImplemented: not implemented for directed type

**Reply by huonw** [reply]
Ah, okay, yes, `EdgeSplitter` doesn't support directed graphs for `keep_connected=True`. If having both of these is important, an alternative would be to implement your own version of edge splitting: the simple case of `method="global"` is relatively straight-forward (although a chunk of code to manage), using random sampling and [`sklearn.model_selection.train_test_split`](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html). If you don't need both of these, you can either switch to an undirected graph (for instance, `StellarGraph.from_networkx(graph.to_undirected(), ...)`) or remove the `keep_connected=True` requirement (or both).

If you wish to go with the connected-and-directed route, https://web.archive.org/web/20200715071324/https://community.stellargraph.io/t/hinsage-link-prediction-and-imbalanced-classification/136 contains some discussion of an approach along these lines (search for `random.choice` in comment `#4` and `model_selection` in comment `#9`) as a starting point. For keeping the graph connected, you could either computed an minimum spanning tree on the undirected graph (and work out the appropriate direction for those selected edges) or use something like [`minimum_spanning_arborescence`](https://networkx.github.io/documentation/stable/reference/algorithms/generated/networkx.algorithms.tree.branchings.minimum_spanning_arborescence.html#networkx.algorithms.tree.branchings.minimum_spanning_arborescence) (I believe).

**Reply by CalmLogarithm** [reply]
Thanks. I'll evaluate the performance without `keep_connected`. It does run now. If I need to I'll come back and try implementing the edge splitting code.
